### PR TITLE
NEW FEATURE: Add dhcid DNS resource type

### PR DESCRIPTION
**Pull request by fritterhoff**
I just stumbled over the missing type dhcid and added it as a first draft. Prop. not that relevant but caused some mini issues migrating existing zones.

### CODE REVIEW

**Comment by fritterhoff** [comment]
I'm not sure if my current draft is fine for a first integration? In theory some further checks could be added but I assume, this RR type is a real corner case in dnscontrol?

**Comment by tlimoncelli** [comment]
(Sorry for the long delay. I was on vacation)

This looks good.  Please rebase and I'll merge.  (I see it is marked "draft"... is there more you'd like to do?)

**Comment by fritterhoff** [comment]
Hi @tlimoncelli,
it is marked as draft since I was unsure if I updated all files and fulfilled all requirements ;-)

File: build/generate/featureMatrix.go [hunk 1]
```diff
@@ -212,6 +214,10 @@ func matrixData() *FeatureMatrix {
 			DomainModifierTlsa,
 			providers.CanUseTLSA,
 		)
+		setCapability(
```
Comment (tlimoncelli) [inline build/generate/featureMatrix.go:217]: Please keep this list alphabetized. (and if anything else is out of place, also move it)

File: models/dnsrr.go [hunk 2]
```diff
@@ -96,6 +96,8 @@ func RRtoRC(rr dns.RR, origin string) (RecordConfig, error) {
 		err = rc.SetTargetTLSA(v.Usage, v.Selector, v.MatchingType, v.Certificate)
 	case *dns.TXT:
 		err = rc.SetTargetTXTs(v.Txt)
+	case *dns.DHCID:
```
Comment (tlimoncelli) [inline models/dnsrr.go:99]: Please keep this list alphabetized. (and if anything else is out of place, also move it)

File: models/record.go [hunk 3]
```diff
@@ -446,6 +446,8 @@ func (rc *RecordConfig) ToRR() dns.RR {
 		rr.(*dns.TLSA).Certificate = rc.GetTargetField()
 	case dns.TypeTXT:
 		rr.(*dns.TXT).Txt = rc.TxtStrings
+	case dns.TypeDHCID:
+		rr.(*dns.DHCID).Digest = rc.GetTargetField()
```
Comment (tlimoncelli) [inline models/record.go:450]: Please keep this list alphabetized. (and if anything else is out of place, also move it)

File: pkg/normalize/validate.go [hunk 4]
```diff
@@ -696,6 +697,7 @@ var providerCapabilityChecks = []pairTypeCapability{
 	capabilityCheck("SRV", providers.CanUseSRV),
 	capabilityCheck("SSHFP", providers.CanUseSSHFP),
 	capabilityCheck("TLSA", providers.CanUseTLSA),
+	capabilityCheck("DHCID", providers.CanUseDHCID),
```
Comment (tlimoncelli) [inline pkg/normalize/validate.go:700]: Please keep this list alphabetized. (and if anything else is out of place, also move it)

File: providers/bind/bindProvider.go [hunk 5]
```diff
@@ -44,6 +44,7 @@ var features = providers.DocumentationNotes{
 	providers.CanUseSRV:              providers.Can(),
 	providers.CanUseSSHFP:            providers.Can(),
 	providers.CanUseTLSA:             providers.Can(),
+	providers.CanUseDHCID:            providers.Can(),
```
Comment (tlimoncelli) [inline providers/bind/bindProvider.go:47]: Please keep this list alphabetized.   (`CanUseDHCID` should be after `CanUseCAA`)

File: providers/capabilities.go [hunk 6]
```diff
@@ -78,6 +78,9 @@ const (
 
 	// DocOfficiallySupported means it is actively used and maintained by stack exchange
 	DocOfficiallySupported
+
+	// CanUseDHCID indicates the provider can handle DHCID records
+	CanUseDHCID
```
Comment (tlimoncelli) [inline providers/capabilities.go:83]: Please keep this list alphabetized.  This item should be after `CanUseCAA`

**Comment by fritterhoff** [comment]
@tlimoncelli fixed all incorrect ordered entries.

**Comment by fritterhoff** [comment]
Ping :)

**Comment by tlimoncelli** [comment]
Looks good so far!    

I went through `documentation/adding-new-rtypes.md` and saw a few things were missing.

1. There are a few places where a case statement needs DHCID added.  Search for `#rtype_variations`:

commands/getZones.go
models/domain.go
models/t_parse.go
models/target.go GetTargetDebug()
pkg/normalize/validate.go

2.  It needs an integration test (integrationTest/integration_test.go).  Look at `testgroup("PTR",` as an example.

3.  It needs documentation.  Look at functions/domain/PTR.md as an example.  (documentation/styleguide-doc.md explains how to add something to the TOC)

Thanks,
Tom

**Comment by fritterhoff** [comment]
Hope I added finally all missing changes

**Comment by tlimoncelli** [comment]
Thanks!  

(I merged this to master outside of github.com by mistake.  Hopefully that didn't mess up the history too badly)

**Comment by tlimoncelli** [comment]
I don't know what I did in git but I've messed things up in ways that I can't fix because... git.

I think I prematurely merged your PR.  Here's a GHA run that shows the problems that hadn't been fixed yet:
https://github.com/StackExchange/dnscontrol/actions/runs/6188855270/job/16802106763

Could you start a new PR?

Apologies for the git mess.

**Comment by fritterhoff** [comment]
No problem. I will fix it tomorrow.

**Pull request merged**